### ISSUE TITLE
Relation to `ase.db`

### ISSUE THREAD

**Issue by jameskermode** [RSE-Cambridge/abcd#3]
How do you see `abcd` as being positioned relative to `ase.db`, which can also use PostgreSQL internally? See https://wiki.fysik.dtu.dk/ase/ase/db/db.html. This already has quite advanced command line query support, so replicating this may not be that worthwhile if it could be imported instead. The planned peer-to-peer push/pull functionality would be new.

**Comment by gabor1** [comment]
The other new idea here is the "discovery" mode, where the query is used as a filter, and rather than returning configurations, it tells you about the statistics of the keys and the values of the configurations that match the filter. this helps you to discover what is in the database,  helps to refine the query

I had a look ase db. yes, we are duplicating some functionality, but really the only one is "take these queries, and pass them straight on as an sql query". so question is whether it would make sense to do what we want (sharing, discovery) on top of existing ase. I am bothered by having to accept all of their "special casing", special names, etc. At the moment I don't feel that the amount of duplication is so much that it is not worth it to be free of those constraints. But I would like to debate this.

**Comment by jameskermode** [comment]
The `ase db` command line tool is quite good for discovery too, it prints the number of matches and shows the top 20 results (by default), e.g.

```
ase convert gp_iter6_sparse9k.xml.xyz GAP6.db
ase db GAP6.db "natoms>40"
ase db GAP6.db -w
```

where the last command opens a nice web-browser view.

I agree the lack of flexibility in key names is the biggest problem, but if we come up with a concrete proposal on importing/exporting general key/value pairs to/from the Atoms.info dict I think there would be willingness to change this.

**Comment by gabor1** [comment]
I tried to import one of our databases, full of properties and parameters, and ase db appeared not  to save any of the info and array variables. 

-- Gábor

**Comment by jameskermode** [comment]
It can, but not by default - and I think not from command line currently The script below add key/value pairs for the per-frame parameters, but needed a filter as it only work for scalars. Not sure about extra arrays either - but my point is that all these things could be fixed in `ase.db` if we wanted, for less effort than a re-write.

```python
import ase.db, ase.io
db = ase.db.connect('abc.db')
for atoms in ase.io.read("infile.xyz@:"):
     info = {k:v for k, v in atoms.info.items() if not hasattr(v, '__len__')}
     db.write(atoms, **info)
```

**Comment by gabor1** [comment]
So all of this (except for the cell :) is already done in abcd, because the storage is via the new JSONB type. Noam wants to store perhaps binary data, input and output files etc, so I think if all of that just works seamlessly and *efficiently* via the JSONB thing, then abcd is already a lot more powerful, and I think has enough advantages over asedb. On the other hand if we find that no, actually for efficiency it really has to be named columns for each property, I agree with you that extending asedb is probably the right thing to do. 

-- Gábor

Gábor Csányi
Professor of Molecular Modelling
Engineering Laboratory
Pembroke College
University of Cambridge

Pembroke College supports CARA. A Lifeline to Academics at Risk.  http://www.cara.ngo/

**Comment by jameskermode** [comment]
Here's a version that includes array parameters and properties too, but not as first-class database columns. 

```python
import ase.db, ase.io
db = ase.db.connect('abc.db')
for atoms in ase.io.read("infile.xyz@:"):
     info = {k:v for k, v in atoms.info.items() if not hasattr(v, '__len__')}
     data = {k:v for k, v in atoms.info.items() if hasattr(v, '__len__')} 
     arrays = {k:v for k, v in atoms.arrays.items() if k not in ['numbers', 'positions']}
     data.update(arrays)
     db.write(atoms, data=data, **info)
```

**Comment by jameskermode** [comment]
Do try the web browser thing, it's really quite nice

**Comment by gabor1** [comment]
Yes, I tried it, it is. And it’s true that adding columns in the browser is tied to those being explicit sql columns (rather than JSONB keys) 

-- Gábor

Gábor Csányi
Professor of Molecular Modelling
Engineering Laboratory
Pembroke College
University of Cambridge

Pembroke College supports CARA. A Lifeline to Academics at Risk.  http://www.cara.ngo/
